src/custom-apps/components/PermissionAlert/utils.ts

```diff
@@ -0,0 +1,145 @@
+  query PermissionIntrospection {
+    __schema {
+      types {
+        kind
+        name
+        description
+        fields(includeDeprecated: false) {
+          name
+          description
+        }
+      }
+    }
+  }
+`;
+
+const uniq = <T>(value: T, index: number, self: T[]) =>
+  self.indexOf(value) === index;
+
+const extractPermissions = (description?: string) => {
+  const match = (description || "").match(/following permissions(.*): (.*?)\./);
+  const permissions = match ? match[2].split(",") : [];
+
+  return permissions;
+};
+
+export const getPermissions = (
+  query: string,
+  permissionMapping: PermissionMap,
+) => {
+  const cursors = extractCursorsFromQuery(query);
+  return cursors.map(findPermission(permissionMapping)).flat().filter(uniq);
+};
+
+export const buildPermissionMap = (
+  elements: IntrospectionNode[],
+): PermissionMap =>
+  elements
+    .filter(({ kind }) => kind === "OBJECT")
+    .filter(
+      ({ name }) =>
+        !/(Created|Create|Delete|Deleted|Update|Updated)$/.test(name),
+    )
+    .reduce((saved, { name, description, fields }) => {
+      const permissions = extractPermissions(description);
+
+      const children = fields.reduce((prev, { name, description }) => {
+        const permissions = extractPermissions(description);
+
+        return {
+          ...prev,
+          [name.toLowerCase()]: { permissions },
+        };
+      }, {});
+
+      return {
+        ...saved,
+        [name.toLowerCase()]: {
+          permissions,
+          children,
+        },
+      };
+    }, {});
+
+const byKind = (name: string) => (element: SelectionNode) =>
+  element.kind === name;
+const extractValue = (element: FieldNode) => element.name.value;
+const isNotEvent = (value: string) => value !== "event";
+
+export const extractCursorsFromQuery = (query: string) => {
+  const cursors = [] as string[][];
+
+  try {
+    const ast = parse(query);
+
+    visit(ast, {
+      Field(node, _key, _parent, _path, ancestors) {
+        if (node.name.value !== "__typename") {
+          const cursor = ancestors
+            .filter(byKind("Field"))
+            .map(extractValue)
+            .filter(isNotEvent);
+
+          if (cursor.length > 0) {
+            cursors.push([...cursor, node.name.value]);
+          }
+        }
+      },
+    });
+  } catch (error) {
+    // explicit silent
+  }
+
+  return cursors;
+};
+
+const groupBy = <T>(list: T[], groupSize = 2) =>
+  list
+    .slice(groupSize - 1)
+    .map((_, index) => list.slice(index, index + groupSize));
+
+export const findPermission =
+  (permissions: PermissionMap) => (cursor: string[]) => {
+    const groups = groupBy(["query", ...cursor]);
+
+    try {
+      return groups.reduce(
+        (saved, [parent, child]) => [
+          ...saved,
+          ...(permissions[parent]
+            ? permissions[parent].children[child].permissions
+            : []),
+        ],
+        [],
+      );
+    } catch {
+      return [];
+    }
```

Comment:
**question**: why is try catch blocked used here?

Reply:
@Droniu fixed and simplified